### PR TITLE
Improve Input Manager error and warning messages

### LINKED ISSUE #635: Make InputManager error messages clearer
When Input Manager is checking the validity of the input data, it would be helpful to note the specific way in which the data failed validation and if the data is not able to be fixed, also note the specific reason why the data was not able to be fixed.

For this issue, error messages in Input Manager `_validate_data()`, `_validate()`, and `_fix_data` should be reviewed and made to be more specific to the errors coming up with the data.

More specificity in the error messages will help the user/developer address the issue.

For V1, it will suffice to provide just a note of which pieces of data weren't valid and whether or not they were able to be fixed.

### PR DESCRIPTION
**Pull request by Isgotkowitz**
This PR improves the error messages provided to the Output Manager by the Input Manager during the input validation and fixing process. The improvements mainly take the form of more specifically stating what is invalid about the data, which data is invalid, and which properties it is violating specifically.

The amount of improvements that can be made to warning and error messages is nearly limitless; this PR seeks to improve messaging while disturbing the interfaces of methods as little as possible. It is highly encouraged that reviewers pull this branch to their local machine and experiment with breaking different inputs to see the message produced.

Closes #635

Examples of warning and error messages:
image

image

image

### CODE REVIEW

**Review by allisterakun — APPROVED** [review]
Great Work!!! LGTM

File: RUFAS/input_manager.py [hunk 1]
```diff
@@ -436,117 +441,163 @@ def _validate_json_element(self, element_hierarchy: List[str], properties_blob_k
...
                 return element_counter_and_validity
 
-    def _array_type_validator(self, variable_properties: Dict[str, Any], var_name: str, input_data_value: list) -> bool:
+    def _array_type_validator(self, variable_properties: Dict[str, Any], var_name: str, input_data_value: list,
+                              properties_blob_key: str) -> bool:
         """Validates an input data element of type array."""
         info_map = {"class": self.__class__.__name__,
                     "function": self._array_type_validator.__name__,
                     }
+        properties_location_string = f" Violates properties defined in '{properties_blob_key}'."
```
Comment (PooyaHekmati) [inline RUFAS/input_manager.py]: can we rename `properties_location_string` to something more representative? The content of this string is different from a location

File: RUFAS/input_manager.py [hunk 2]
```diff
@@ -436,117 +441,163 @@ def _validate_json_element(self, element_hierarchy: List[str], properties_blob_k
...
         if type(input_data_value) is not list:
             warning_string = "Validation: array is not a list."
-            om.add_warning(warning_string, f"{var_name=}", info_map)
+            warning_message = f"Variable '{var_name}' is type: {type(input_data_value)}"
+            warning_message += properties_location_string
```
Comment (PooyaHekmati) [inline RUFAS/input_manager.py]:
```suggestion
            warning_message = f"Variable '{var_name}' is type: {type(input_data_value)}{properties_location_string}"
```

File: RUFAS/input_manager.py (hunk 1 above)
Comment (Isgotkowitz) [reply RUFAS/input_manager.py]: Yes. I went with `properties_violation_message`, but can easily change again as needed.

File: RUFAS/input_manager.py (hunk 2 above)
Comment (Isgotkowitz) [reply RUFAS/input_manager.py]: Implemented

Comment (PooyaHekmati) [reply RUFAS/input_manager.py]: I still `warning_message += properties_location_string` here and 10 other occurrences.

File: RUFAS/input_manager.py (hunk 1 above)
Comment (PooyaHekmati) [reply RUFAS/input_manager.py]: i still see `properties_location_string`

Comment (Isgotkowitz) [reply RUFAS/input_manager.py]: @PooyaHekmati my bad, forgot to push my changes.

**Review by PooyaHekmati — APPROVED** [review]

File: RUFAS/input_manager.py [hunk 3]
```diff
@@ -436,117 +441,154 @@ def _validate_json_element(self, element_hierarchy: List[str], properties_blob_k
...
         return True
 
-    def _string_type_validator(self, variable_properties: Dict[str, Any], var_name: str, input_data_value: str) -> bool:
+    def _string_type_validator(self, variable_properties: Dict[str, Any], var_name: str, input_data_value: str,
+                               properties_blob_key: str) -> bool:
         """Validates an input data string element."""
         info_map = {"class": self.__class__.__name__,
                     "function": self._string_type_validator.__name__,
                     }
+        properties_violation_message = f"Violates properties defined in '{properties_blob_key}'."
```
Comment (ew3361zh) [inline RUFAS/input_manager.py]: Is the idea that people would know where to find the metadata and would go to look in the metadata and then correct their input? If so, think I'd be more specific that the properties section they're looking for is in the metadata:

```suggestion
        properties_violation_message = f"Violates properties defined in metadata '{properties_blob_key}'."
```

**Comment by ew3361zh** [comment]
Small suggestion here as well but could we change the punctuation of the output message? It looks a little clunky right now with periods before semicolons:
Screenshot 2023-11-16 at 9 23 02 AM

So for example this:
[16-Nov-2023_Thu_09-07-47.637704][WARNING][default_scenario] Validation: value is not a number.: Variable 'replace_num' has value: fifty, is type: <class 'str'>. Violates properties defined in 'animal_properties'.

I think would look better as:
[16-Nov-2023_Thu_09-07-47.637704][WARNING][default_scenario] Validation: value is not a number. Variable: 'replace_num' has value: fifty, is type: <class 'str'>. Violates properties defined in 'animal_properties'.

Which would mean updating the warning message text in the various type-validator functions from "Variable '{var_name}'..." to "Variable: '{var_name}'..."

And then in OM changing the log_format from:
log_format = "{color}[{timestamp}][{log_level}][{metadata_prefix}] {name}: {message}{color_reset}\n"

to

log_format = "{color}[{timestamp}][{log_level}][{metadata_prefix}] {name} {message}{color_reset}\n"

**Comment by Isgotkowitz** [comment]
@ew3361zh good suggestions, thank you.

File: RUFAS/input_manager.py (hunk 3 above)
Comment (Isgotkowitz) [reply RUFAS/input_manager.py]: Yes, that is generally the idea. Ideally, the messages would specify the top level blob of data in `files` that contains that data in violation of the properties. But making that happen would require some pretty significant changes so I left them out of this PR.

File: RUFAS/input_manager.py [hunk 4]
```diff
@@ -234,7 +234,8 @@ def _filter_input_data_by_metadata(self, input_data: Dict[str, Any],
 
         return filtered_input_data
 
-    def _validate_input_type_dynamic(self, variable_properties: Dict[str, Any], var_name: str, input_data_value: Any):
+    def _validate_input_type_dynamic(self, variable_properties: Dict[str, Any], var_name: str, input_data_value: Any,
+                                     properties_blob_key: str) -> bool:
```
Comment (krb78) [inline RUFAS/input_manager.py:238]:
```suggestion
                                     properties_blob_key: str) -> bool:
```

**Review by krb78 — COMMENTED** [review]
PR #1 for me :D

**Pull request merged**